### PR TITLE
AMQP-581: Expose Caching CF Cache Statistics

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/AMQP-581

### CODE REVIEW

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java [hunk 1]
```diff
@@ -676,6 +681,44 @@ protected void reset(List<ChannelProxy> channels, List<ChannelProxy> txChannels)
 		this.connection = null;
 	}
 
+	@ManagedAttribute
+	public Properties getCacheProperties() {
+		Properties props = new Properties();
+		props.setProperty("cachMode", this.cacheMode.name());
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: Typo: cach**e**Mode

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java [hunk 2]
```diff
@@ -676,6 +681,44 @@ protected void reset(List<ChannelProxy> channels, List<ChannelProxy> txChannels)
...
+		synchronized(connectionMonitor) {
+			props.setProperty("channelCacheSize", Integer.toString(this.channelCacheSize));
+			if (this.cacheMode.equals(CacheMode.CONNECTION)) {
+				props.setProperty("connectionCacheSize", Integer.toString(this.connectionCacheSize));
+				props.setProperty("openConnections", Integer.toString(this.openConnections.size()));
+				props.setProperty("idleConnections", Integer.toString(this.idleConnections.size()));
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: Would you mind to explain the reason just don't use `"" + this.connectionCacheSize`?

I intend to believe the comment on SO (http://stackoverflow.com/questions/14126792/why-should-i-use-integer-tostring-instead-of-just-printing-the-integer):
> For string concatination expressions like
> `"The number is " + k;`
> javac builds the actual code using `StringBuilder` as
> `new StringBuilder("The number is ")).append(Integer.toString(k)).toString());`

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java [hunk 3]
```diff
@@ -676,6 +681,44 @@ protected void reset(List<ChannelProxy> channels, List<ChannelProxy> txChannels)
...
+				int n = 0;
+				for (Entry<ChannelCachingConnectionProxy, LinkedList<ChannelProxy>> entry :
+										this.openConnectionTransactionalChannels.entrySet()) {
+					int port = entry.getKey().getLocalPort();
+					if (port == 0) {
+						port = ++n; // just use a unique id to avoid overwriting
+					}
+					props.put("idleChannelsTx:" + port, Integer.valueOf(entry.getValue().size()));
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: unnecessary_boxing

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java (hunk 2 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: Saves an object creation (`StringBuilder`) and I think `"" + foo` looks clumsy.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java (hunk 3 above)
Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: Yup - unneeded.

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java [hunk 4]
```diff
@@ -676,6 +681,44 @@ protected void reset(List<ChannelProxy> channels, List<ChannelProxy> txChannels)
...
+				}
+				for (Entry<ChannelCachingConnectionProxy, LinkedList<ChannelProxy>> entry :
+										this.openConnectionNonTransactionalChannels.entrySet()) {
+					int port = entry.getKey().getLocalPort();
+					if (port == 0) {
+						port = ++n; // just use a unique id to avoid overwriting
+					}
+					props.put("idleChannelsNotTx:" + port, Integer.valueOf(entry.getValue().size()));
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: DITO

**Comment by garyrussell** [comment]
Pushed

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java [hunk 5]
```diff
@@ -676,6 +681,44 @@ protected void reset(List<ChannelProxy> channels, List<ChannelProxy> txChannels)
...
+						port = ++n; // just use a unique id to avoid overwriting
+					}
+					props.put("idleChannelsNotTx:" + port, entry.getValue().size());
+				}
+			}
+			else {
+				props.setProperty("localPort",
+						Integer.toString(this.connection == null ? 0 : this.connection.getLocalPort()));
+				props.setProperty("idleChannelsTx", Integer.toString(cachedChannelsTransactional.size()));
+				props.setProperty("idleChannelsNotTx", Integer.toString(cachedChannelsNonTransactional.size()));
```
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: I will fix all `this.` vulnerabilities on merge.

Comment (garyrussell) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/connection/CachingConnectionFactory.java]: thanks

File: spring-rabbit/src/test/java/org/springframework/amqp/rabbit/connection/CachePropertiesTests.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		Properties props = this.connectionCf.getCacheProperties();
+		assertEquals("10", props.getProperty("channelCacheSize"));
+		assertEquals("5", props.getProperty("connectionCacheSize"));
+		assertEquals("2", props.getProperty("openConnections"));
+		assertEquals("1", props.getProperty("idleConnections"));
+		c2.close();
+	}
+
+	@Configuration
+	@EnableMBeanExport
```
Comment (artembilan) [inline spring-rabbit/src/test/java/org/springframework/amqp/rabbit/connection/CachePropertiesTests.java]: I guess this is redundant for test, but we need to to get that cool picture for docs from JMX :smile:

**Comment by garyrussell** [comment]
Repushed with fixed image.

**Comment by artembilan** [comment]
LGTM. 
Merging...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-amqp/commit/6da60b7b4ca4441623b920e507ea0a6b0248a7f6

**Pull request closed without merging**